vm/tests/runner.rs

```diff
@@ -0,0 +1,283 @@
+use compiler::{compile, CompilerOptions};
+use context::source::Source;
+use parser::parse_trusted;
+use vm::value::VmValue;
+use vm::{VmError, VmValueFFI, VM};
+
+pub struct Runner<'a> {
+    externals: Externals<'a>,
+    vm: VM,
+    analyzer: Analyzer<'a>,
+    current_page: Option<usize>,
+}
+
+impl Default for Runner<'_> {
+    fn default() -> Self {
+        let mut externals = Externals::default();
+        let mut std_importer = FileImporter::new(PathBuf::from("../lib"));
+        let mut vm = VM::default();
+
+        let std_name = Name::new("std");
+        let analyzer = analyze(std_name.clone(), &mut std_importer, &externals);
+        let mut buff = Vec::new();
+
+        compile(
+            &analyzer.engine,
+            &analyzer.typing,
+            &analyzer.resolution.relations,
+            &analyzer.resolution.engine,
+            &externals,
+            externals.current,
+            SourceId(0),
+            &mut buff,
+            CompilerOptions::default(),
+        )
+        .expect("std did not compile successfully");
+
+        vm.register(&buff).expect("VM std register");
+        unsafe {
+            vm.run().expect("VM std init");
+        }
+
+        externals.register(Reef::new("std".to_string(), analyzer));
+
+        Self {
+            externals,
+            vm,
+            analyzer: Analyzer::default(),
+            current_page: None,
+        }
+    }
+}
+
+impl<'a> Runner<'a> {
+    pub fn eval(&mut self, expr: &'a str) -> VmValue {
+        match self.try_eval(expr) {
+            Ok(v) => v,
+            Err(VmError::Panic) => panic!("VM did panic"),
+            Err(VmError::Internal) => panic!("VM internal error"),
+        }
+    }
+
+    pub fn try_eval(&mut self, expr: &'a str) -> Result<VmValue, VmError> {
+        let name = Name::new("runner");
+        let src = Source::unknown(expr);
+        let mut importer = StaticImporter::new([(name.clone(), src)], parse_trusted);
+        let ImportResult::Success(imported) = importer.import(&name) else {
+            unreachable!()
+        };
+
+        let inject = Inject {
+            name: name.clone(),
+            imported,
+            attached: self.current_page.map(SourceId),
+        };
+
+        let mut analysis = self.analyzer.inject(inject, &mut importer, &self.externals);
+        let page = analysis.attributed_id();
+        self.current_page = Some(page.0);
+        let diagnostics = analysis.take_diagnostics();
+
+        let reef = self.externals.current;
+
+        if !diagnostics.is_empty() {
+            panic!("input had analysis errors")
+        }
+        let mut bytes = Vec::new();
+
+        let chunk = self.analyzer.engine.get_user(page).unwrap();
+        let chunk_expr = &chunk.expression.as_ref().unwrap();
+
+        let evaluated_expr_type = if let ExprKind::Block(exprs) = &chunk_expr.kind {
+            exprs.last().map_or(types::UNIT, |v| v.ty)
+        } else {
+            chunk_expr.ty
+        };
```

Comment:
```suggestion
        let evaluated_expr_type = chunk_expr.ty;
```
The typing phase already does that, recursively.